dashboard/app/helpers/aichat_ai_client.rb

```diff
@@ -0,0 +1,270 @@
+# Combined class to hold shared logic for multiple AI API backends. This is essentially an "abstract" class
+# that is never instantiated directly. The derived classes hold implementation details in required overridden
+# methods. Currently the two implemented APIs (OpenAI and Gemini) are POST based REST APIs
+class AichatAiClient
+  # get an instance of the appropriate derived class based on model id
+  def self.get_instance(model_id)
+    # TODO - we should create a map/hash from model_id to model/key (not sure why model id isn't just model but whatever)
+    # but for now we just assume it's one of the gemini models if not 'gpt-4o-mini'
+    if model_id == "gpt-4o-mini"
+      return AichatOpenaiCompletionsClient.new(CDO.openai_student_learning_api_key, SharedConstants::AICHAT_MODEL_VERSION)
+    else
+      return AichatGeminiClient.new(CDO.google_gemini_student_learning_api_key, model_id)
+    end
+  end
+
+  # call the API (through methods overridden in derived class) and get response text to send back to user
+  def get_response_text(stored_messages, new_message, temperature, system_prompt, retrieval_contexts,  model_id, level_id, encrypted_channel_id, user_id, project_id)
+    start_time = Time.now
+
+    level = Level.find_by(id: level_id)
+
+    # level system prompt - string or nil
+    # TODO - determine if we're still using level system prompts
+    level_system_prompt = level&.properties&.dig('aichat_settings', 'levelSystemPrompt')
+
+    # level name - string
+    level_name = level&.name
+
+    system_instructions = combine_system_instructions(
+      system_prompt,
+      level_system_prompt,
+      retrieval_contexts
+    )
+
+    body = create_body(
+      stored_messages,
+      new_message,
+      system_instructions,
+      temperature,
+      level_name,
+      encrypted_channel_id
+    )
+
+    http_response = HTTParty.post(
+      url,
+      headers: headers,
+      body: body.to_json,
+      open_timeout: DCDO.get('openai_http_open_timeout', 5),
+      read_timeout: DCDO.get('openai_http_read_timeout', 30)
+    )
+
+    response_body = JSON.parse(http_response.body)
+
+    raise_possible_response_errors_from_body(response_body)
+
+    # TODO - the fact we called this extract_text_response_from_body suggests we need to adapt this class for multimodal
+    # *responses* even if we have a way to disable them in level builder
```

Comment:
We don't currently have any support for multimodal/non-text responses (only inputs)

Reply:
Yep sure! Agreed that we'd likely want to update the function name if/when we support multimodal responses, not to mention all the other work to support passing that binary data back to the user 🙂

Reply:
nit: mind removing/updating this TODO? Assuming this is referring to multimodal responses, we don't currently have "a way to disable them in levelbuilder" (this could instead be something like "update the function name when supporting multimodal responses" or similar?)